examples/fluids/navierstokes.h

```diff
@@ -20,6 +20,10 @@
 #error "PETSc v3.20 or later is required"
 #endif
 
+#if ((PETSC_VERSION_MAJOR < 4) && (PETSC_VERSION_MINOR < 20))
```

Comment:
This should be enough. The change is not present in 3.20 release.
```suggestion
#if PETSC_VERSION_LT(3, 21, 0)
```

Reply:
Perhaps I am misunderstanding something so I will blog out what I see:
```
(base) kjansen@viz003: /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork ((59d2de5a06b...))$ export  PETSC_ARCH=64DebugUpstream
(base) kjansen@viz003: /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork ((59d2de5a06b...))$ grep PETSC_VERSION configure.log
          PETSC_VERSION_RELEASE of 1 indicates the code is from a release branch or a branch created from a release branch.
          #define PETSC_VERSION_RELEASE    0
          #define PETSC_VERSION_MAJOR      3
          #define PETSC_VERSION_MINOR      20
          #define PETSC_VERSION_SUBMINOR   1
          #define PETSC_VERSION_DATE       "unknown"
          #define PETSC_VERSION_GIT        "unknown"
          #define PETSC_VERSION_DATE_GIT   "unknown"
          #define PETSC_VERSION_EQ(MAJOR,MINOR,SUBMINOR) \
          #define PETSC_VERSION_ PETSC_VERSION_EQ
          #define PETSC_VERSION_LT(MAJOR,MINOR,SUBMINOR)          \
          #define PETSC_VERSION_LE(MAJOR,MINOR,SUBMINOR) \
          #define PETSC_VERSION_GT(MAJOR,MINOR,SUBMINOR) \
          #define PETSC_VERSION_GE(MAJOR,MINOR,SUBMINOR) \
#define PETSC_VERSION_BRANCH_GIT "HEAD"
#define PETSC_VERSION_DATE_GIT "2023-11-03 14:14:42 +0000"
#define PETSC_VERSION_GIT "v3.20.1-137-g59d2de5a06b"
(base) kjansen@viz003: /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork ((59d2de5a06b...))$ git log |head -10
commit 59d2de5a06b82e860b55768cbaf4b28f638fcdc8
Merge: 9baa572492f 3e6b6c1625a
Author: Satish Balay <balay@mcs.anl.gov>
Date:   Fri Nov 3 14:14:42 2023 +0000

    Merge branch 'jolivet/update-openmpi' into 'main'
    
    configure: OpenMPI 5.0.0
    
    See merge request petsc/petsc!6971
```

That is a today pull of main.  
With this, [4897600](https://github.com/CEED/libCEED/pull/1391/commits/4897600808f796708ed95be171823b7eccf8af7a)
works.  That commit also works with my older builds on my September 18, 2023 fork. However, if I make the change you suggest, as you can see below, I get the warnings shown  AND it won't compile in CI (which is also using a 3.19 PETSC) 

```
base) kjansen@viz003: /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork (kjansen/AddingPeriodicSF)$ grep PETSC_VERSION 64Debughdf5AndPartitioners/lib/petsc/conf/configure.log
          PETSC_VERSION_RELEASE of 1 indicates the code is from a release branch or a branch created from a release branch.
          #define PETSC_VERSION_RELEASE    0
          #define PETSC_VERSION_MAJOR      3
          #define PETSC_VERSION_MINOR      19
          #define PETSC_VERSION_SUBMINOR   5
          #define PETSC_VERSION_DATE       "unknown"
          #define PETSC_VERSION_GIT        "unknown"
          #define PETSC_VERSION_DATE_GIT   "unknown"
          #define PETSC_VERSION_EQ(MAJOR,MINOR,SUBMINOR) \
          #define PETSC_VERSION_ PETSC_VERSION_EQ
          #define PETSC_VERSION_LT(MAJOR,MINOR,SUBMINOR)          \
          #define PETSC_VERSION_LE(MAJOR,MINOR,SUBMINOR) \
          #define PETSC_VERSION_GT(MAJOR,MINOR,SUBMINOR) \
          #define PETSC_VERSION_GE(MAJOR,MINOR,SUBMINOR) \
#define PETSC_VERSION_BRANCH_GIT "kjansen/AddingPeriodicSF"
#define PETSC_VERSION_DATE_GIT "2023-11-03 09:47:57 -0600"
#define PETSC_VERSION_GIT "v3.19.5-1184-g250ea0c9bab"
(base) kjansen@viz003: /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork (kjansen/AddingPeriodicSF)$ cd -
/projects/tools/libCEED/libCEED/examples/fluids
(base) kjansen@viz003: /projects/tools/libCEED/libCEED/examples/fluids (kjansen/ReadCGNS)$ make clean
rm -f -r build navierstokes *.vtu *.bin* *.csv *.png
(base) kjansen@viz003: /projects/tools/libCEED/libCEED/examples/fluids (kjansen/ReadCGNS)$ make -j
          CC build/navierstokes.o
          CC build/problems/advection.o
          CC build/problems/advection2d.o
          CC build/problems/blasius.o
          CC build/problems/channel.o
          CC build/problems/densitycurrent.o
          CC build/problems/eulervortex.o
          CC build/problems/freestream_bc.o
          CC build/problems/gaussianwave.o
          CC build/problems/newtonian.o
          CC build/problems/sgs_dd_model.o
          CC build/problems/shocktube.o
          CC build/problems/stg_shur14.o
          CC build/problems/taylorgreen.o
          CC build/src/cloptions.o
          CC build/src/differential_filter.o
          CC build/src/dm_utils.o
          CC build/src/grid_anisotropy_tensor.o
          CC build/src/misc.o
          CC build/src/petsc_ops.o
          CC build/src/setupdm.o
          CC build/src/setuplibceed.o
          CC build/src/setupts.o
          CC build/src/strong_boundary_conditions.o
          CC build/src/turb_spanstats.o
          CC build/src/velocity_gradient_projection.o
In file included from /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscsys.h:1100,
                 from /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscsection.h:3,
                 from /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscdmplex.h:6,
                 from /projects/tools/libCEED/libCEED/examples/fluids/src/dm_utils.c:12:
/projects/tools/libCEED/libCEED/examples/fluids/src/dm_utils.c: In function \u2018DMSetupByOrderBegin_FEM\u2019:
/projects/tools/libCEED/libCEED/examples/fluids/src/dm_utils.c:460:15: warning: implicit declaration of function \u2018DMSetCoordinateDisc\u2019; did you mean \u2018DMSetCoordinateDim\u2019? [-Wimplicit-function-declaration]
  460 |     PetscCall(DMSetCoordinateDisc(dm, fe_coord_new, PETSC_TRUE));
      |               ^~~~~~~~~~~~~~~~~~~
/projects/tools/libCEED/libCEED/examples/fluids/src/dm_utils.c:460:5: note: in expansion of macro \u2018PetscCall\u2019
  460 |     PetscCall(DMSetCoordinateDisc(dm, fe_coord_new, PETSC_TRUE));
      |     ^~~~~~~~~
In file included from /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscsys.h:1100,
                 from /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscsection.h:3,
                 from /projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscdmplex.h:6,
                 from /projects/tools/libCEED/libCEED/examples/fluids/src/turb_spanstats.c:17:
/projects/tools/libCEED/libCEED/examples/fluids/src/turb_spanstats.c: In function \u2018CreateStatsDM\u2019:
/projects/tools/libCEED/libCEED/examples/fluids/src/turb_spanstats.c:71:15: warning: implicit declaration of function \u2018DMSetCoordinateDisc\u2019; did you mean \u2018DMSetCoordinateDim\u2019? [-Wimplicit-function-declaration]
   71 |     PetscCall(DMSetCoordinateDisc(user->spanstats.dm, NULL, PETSC_TRUE));  // Ensure that a coordinate FE exists
      |               ^~~~~~~~~~~~~~~~~~~
/projects/tools/petsc/petsc_kjansenFork/petsc-kjansen-fork/include/petscerror.h:460:28: note: in definition of macro \u2018PetscCall\u2019
  460 |       ierr_petsc_call_q_ = __VA_ARGS__; \
      |                            ^~~~~~~~~~~
      LINK.o ./navierstokes
rm build/src/.DIR build/problems/.DIR build/.DIR
(base) kjansen@viz003: /projects/tools/libCEED/libCEED/examples/fluids (kjansen/ReadCGNS)$ git diff 
diff --git a/examples/fluids/navierstokes.h b/examples/fluids/navierstokes.h
index 8ed62e9b..5b73bd25 100644
--- a/examples/fluids/navierstokes.h
+++ b/examples/fluids/navierstokes.h
@@ -20,7 +20,8 @@
 #error "PETSc v3.20 or later is required"
 #endif
 
-#if ((PETSC_VERSION_MAJOR < 4) && (PETSC_VERSION_MINOR < 20))
+// #if ((PETSC_VERSION_MAJOR < 4) && (PETSC_VERSION_MINOR < 20))
+#if PETSC_VERSION_LT(3, 21, 0)
 #define DMSetCoordinateDisc(a, b, c) DMProjectCoordinates(a, b)
 #endif
```

I will commit and push the libCEED change which will trigger a CI so that you can see why I switched to using major and minor version instead (which does work for both new and old PETSc).

Reply:
The CI has already failed to build because the compounded version is not doing what we expected/wanted it to do.  The only change from the working version is the return to a compounded version replacing my split separate check of major and minor version.

Reply:
The simpler proof that compounded is not working is the lines above it.  If error is supposed to stop the code, its clear that 3.19 is not stopping on a > 3.20.0 limit.

Reply:
The build in CI is showing `PETSC_VERSION_RELEASE    0`, so it will always read as further ahead of any version number we pick.

The fix is to use the simpler macro and update the CI PETSc build, if I'm understanding correctly.

Reply:
I am not sure if it is any "simpler" but checking major and minor version "works" in the sense that both new (stuff after Matt renamed)  and old PETSc (stuff before) will build.

Reply:
Sure, but the offending commit is in between two releases. There will always be commits versions of PETSc main we cannot support. We historically have rebuilt CI PETSc instead of trying to chase support for stale versions of PETSc main. I'm not sure what trying to support stale, non-release versions of PETSc main buys us?

Reply:
I suspect I am confused  and divergent from the objective somehow.  Maybe this is my issue but I have been in the habit of building against petsc `main` where I find

1) Every version of petsc main that I try after 
```
commit e44f6aeb0c7eee2716d50e70cd976c48a5d3c225
Author: Matthew Knepley <knepley@gmail.com>
Date:   Wed Oct 25 22:52:09 2023 -0400

    Plex: Enable high order viewing - Refinement recreates affine coordinates - Default for draw viewer - Replace DMProjectCoordinates() with DMSetCoordinateDisc() - Add argument to DMPlexCreateCoordinateSpace()
```
has this issue.  I am confused by the statement "chase support for stale versions`.
2) doing a fresh pull of petsc main and searching backwards on release has  10/31 `Increase patch level to 3.20.1`
3) similarly on 9/25 `set petsc v3.20.0`

If I am mistaken or if there a ways to use newer PETSc versions AND still have functional code with versions of PETSc prior to Matt's commit then we can kill  this PR and move forward a different way.

If I am not way off track in my understanding, this is well into bike shedding territory so all I am looking for here is a thumbs up or down on  this PR (after putting back to something that passes CI of course...I have left it in a broken state for @jedbrown to assess).

Reply:
Ah, I think I see the confusion. Historically, it has been our policy when creating  these versioning support macros to only support

1) A tagged release

2) The current most recent commit to main

Other commits to main *might* work fine, but we don't try to make our codes run against every commit to PETSc main. That's hard.

Our PETSc build for this CI pipeline is too stale, so we need to update our build for this pipeline. Then our typical approach for these compatibility macros will work.